src/Nethermind/Nethermind.Era1/E2StoreReader.cs

```diff
@@ -0,0 +1,206 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Buffers;
+using System.Buffers.Binary;
+using System.IO.Compression;
+using System.Security.Cryptography;
+using CommunityToolkit.HighPerformance;
+using DotNetty.Buffers;
+using Microsoft.IO;
+using Microsoft.Win32.SafeHandles;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Resettables;
+using Snappier;
+
+namespace Nethermind.Era1;
+
+public class E2StoreReader : IDisposable
+{
+    private const int HeaderSize = 8;
+    private const int IndexSectionCount = 8;
+    private const int IndexSectionStartBlock = 8;
+    private const int IndexOffsetSize = 8;
+    private const int ValueSizeLimit = 1024 * 1024 * 50;
+
+    private readonly SafeFileHandle _file;
+
+    // Read these two value ahead of time instead of fetching the value everything it is needed to reduce
+    // the page fault when looking up.
+    private long? _startBlock;
+    private long _blockCount;
+    private readonly long _fileLength;
+
+    public E2StoreReader(string filePath)
+    {
+        _file = File.OpenHandle(filePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
+        _fileLength = RandomAccess.GetLength(_file);
+    }
+
+    public (T, long) ReadEntryAndDecode<T>(long position, Func<Memory<byte>, T> decoder, ushort expectedType)
```

Comment:
minor, I think this would be more efficient on stack handling for structs:
```suggestion
    public long ReadEntryAndDecode<T>(long position, Func<Memory<byte>, T> decoder, ushort expectedType, out T value)
```